### PR TITLE
Fixes HFR ui

### LINKED ISSUE #74677: HFR recipe selection is broken in NTOS
Reporting client version: 514.1589

## Round ID:
[203620](https://scrubby.melonmesa.com/round/203620)

## Testmerges:
- [Adds a Material Stack manipulation apparatus for Engineer Cyborgs, replaces the R-Glass tool](https://www.github.com/tgstation/tgstation/pull/74297)
- [Fixes taste not existing for mobs without having their tongue changed](https://www.github.com/tgstation/tgstation/pull/74665)

## Reproduction: Trying to select a fuel fusion recipe in the HFR results in a fatal error bluescreen in NTOS.
1. Construct the HFR
2. Try to select a fuel recipe (IE, something like plasma/oxygen fusion fuel)
3. Get bluescreened

Additionally, refreshing tgui will make it so that it does not crash when you try to select a recipe any more, but it will not open the recipe selection page.

### PR DESCRIPTION
**Pull request by jlsnow301**
## About The Pull Request
Issue is caused by the recipe dropdown inside of the HFR trying to make table cells using strings that COULD be null. 

This fixes other issues present and converts it to typescript, so I didn't add the no gbp label.

The ui now has better margins and the recipe menu now properly selects new recipes.

Fixes #74677

## Why It's Good For The Game
TS good
bug fixes

## Changelog
:cl:
fix: The HFR screen should be fixed, let me know if you see glitches
fix: The HFR screen should now also select recipes properly
/:cl:

### CODE REVIEW

**Comment by orthography** [comment]
Atmos is so esoteric it's not in DM.

File: tgui/packages/tgui/interfaces/Hypertorus/Controls.tsx [hunk 1]
```diff
@@ -203,18 +241,17 @@ export const HypertorusWasteRemove = (props, context) => {
               Filter from moderator mix:
             </>
           }>
-          {filterTypes.map((filter) => (
-            <Button
-              key={filter.gas_id}
-              icon={filter.enabled ? 'check-square-o' : 'square-o'}
-              selected={filter.enabled}
-              content={getGasLabel(filter.gas_id, filter.gas_name)}
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/Hypertorus/Controls.tsx:211]: one error here

File: tgui/packages/tgui/interfaces/Hypertorus/index.tsx [hunk 2]
```diff
@@ -1,51 +1,94 @@
...
       </Stack>
       <Collapsible title="Recipe selection">
         <HypertorusRecipes
-          baseMaximumTemperature={data.base_max_temperature}
-          enableRecipeSelection={data.power_level === 0}
+          baseMaxTemperature={base_max_temperature}
+          enableRecipeSelection={power_level === 0}
           onRecipe={(id) => act('fuel', { mode: id })}
-          selectableFuels={data.selectable_fuel}
-          selectedFuelID={data.selected}
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/Hypertorus/index.tsx:48]: this is another bug. for the record, this is bad. there's no reason (that I know of) that you should take backend data then shove it into props. components can call backend data anywhere, so this was just causing it to not update

File: tgui/packages/tgui/interfaces/Hypertorus/Recipes.tsx [hunk 3]
```diff
@@ -110,25 +121,23 @@ const MemoRow = (props) => {
   );
 };
 
-MemoRow.defaultHooks = activeChange;
-
-const GasCellItem = (props) => {
+const GasCellItem = (props: GasCellProps) => {
   const { gasid, ...rest } = props;
+
+  if (!gasid) return <Table.Cell />;
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/Hypertorus/Recipes.tsx:127]: this fixes the second error. this just fell through before, there would be no id so the table and inner box would return blank table cell. i don't want getGasLabel to get called without a string at all, so denied

File: tgui/packages/tgui/constants.ts [hunk 4]
```diff
@@ -291,35 +291,50 @@ const GASES = [
...
   const gasSearchString = gasId.toLowerCase();
-  const gas = GASES.find(
-    (gas) =>
-      gas.id === gasSearchString || gas.name.toLowerCase() === gasSearchString
-  );
-  return gas?.label || fallbackValue || gasId;
+
+  for (let idx = 0; idx < GASES.length; idx++) {
+    if (GASES[idx].id !== gasSearchString) continue;
+    return GASES[idx].label;
```
Comment (stylemistake) [inline tgui/packages/tgui/constants.ts:300]:
```suggestion
    if (GASES[idx].id === gasSearchString) return GASES[idx].label;
```

File: tgui/packages/tgui/constants.ts [hunk 5]
```diff
@@ -291,35 +291,50 @@ const GASES = [
...
-  const gas = GASES.find(
-    (gas) =>
-      gas.id === gasSearchString || gas.name.toLowerCase() === gasSearchString
-  );
-  return gas?.color;
+
+  for (let idx = 0; idx < GASES.length; idx++) {
+    if (GASES[idx].id !== gasSearchString) continue;
+    return GASES[idx].color;
+  }
```
Comment (stylemistake) [inline tgui/packages/tgui/constants.ts:315]:
```suggestion
  for (let idx = 0; idx < GASES.length; idx++) {
    if (GASES[idx].id === gasSearchString) return GASES[idx].color;
  }
```

**Review by stylemistake — APPROVED** [review]
Looks good, left one comment.

File: tgui/packages/tgui/interfaces/Hypertorus/Recipes.tsx (hunk 3 above)
Comment (stylemistake) [reply tgui/packages/tgui/interfaces/Hypertorus/Recipes.tsx:127]: 👍

**Comment by SyncIt21** [comment]
Can we please get this merged? atmos is no fun without the HFR

**Comment by plutori** [comment]
Merge please... I miss the HFR.

**Pull request merged**